ce/api/streamflow/watershed.py

```diff
@@ -0,0 +1,421 @@
+        flow_direction.values,
+        direction_matrix,
+        debug=True
+    )
+
+    # `watershed_lonlats`, `elevations`, and `areas` must all be ordered
+    # collections (not sets) because it is required (at minimum) that the
+    # coordinates (lonlats) for `elevations[i]` and `areas[i]` be equal for
+    # all `i`.
+    watershed_lonlats = \
+        [flow_direction.xy_to_lonlat(xy) for xy in watershed_xys]
+
+    #  Compute elevations at each lonlat of watershed
+    ws_elevations = elevation.get_values_at_lonlats(watershed_lonlats)
+
+    #  Compute area of each cell in watershed
+    ws_areas = area.get_values_at_lonlats(watershed_lonlats)
+
+    # Compute the elevation/area curve
+    cumulative_areas = hypsometry(ws_elevations, ws_areas, **hypso_params)
+
+    # Compute outline of watershed as a GeoJSON feature
+    outline = outline_cell_rect(
+        watershed_lonlats, flow_direction.lat_step, flow_direction.lon_step)
+
+    return {
+        'elevation': {
+            'units': elevation.units,
+            'minimum': min(ws_elevations),
+            'maximum': max(ws_elevations),
+        },
+        'area': {
+            'units': area.units,
+            'value': sum(ws_areas),
+        },
+        'hypsometric_curve': {
+            'elevation_bin_start': hypso_params['bin_start'],
+            'elevation_bin_width': hypso_params['bin_width'],
+            'elevation_num_bins': hypso_params['num_bins'],
+            'cumulative_areas': cumulative_areas,
+            'elevation_units': elevation.units,
+            'area_units': area.units,
+        },
+        'boundary': geojson_feature(
+            outline,
+            properties={
+                # Represent as GeoJSON?
+                'mouth': {
+                    'longitude': station_lonlat[0],
+                    'latitude': station_lonlat[1]
+                }
+            },
+        ),
+        'debug/test': {
+            'watershed': {
+                'cell_count': len(watershed_xys),
+                **watershed_debug
+            }
+        }
+    }
+
+
+def build_watershed(target, routing, direction_map, debug=False):
+    """
+    Return set of all cells (including target) that drain into `target`.
+
+    :param target: An xy index representing the cell of interest.
+    :param routing: A numpy array representing water flow using VIC direction
+        codes (0 - 9). These codes may be integers or floats.
+    :param direction_map: Maps VIC direction codes into offsets in cell indices.
+        Necessary because, depending on whether lon and lat dimensions
+        increase or decrease with increasing index, a move north or east is
+        represented by an offset of +1 or -1, respectively.
+        TODO: Compute this internally?
+    :param debug: Boolean indicating whether this function should compute
+        and return debug information.
+    :return: Set of cells (cell indices) that drain into `target`.
```

Comment:
Only if we are comfortable with requiring that flow_direction, elevation, and area datasets are on *exactly* the same grid, so that the indices in each for a given latlon are identical. The mapping back and forth makes it robust to differences in grid, which might be a good thing, or might not. See also my question in response to comment on compatibility tests.